ImageFeed/Services/ProfileImageService.swift

```diff
@@ -0,0 +1,89 @@
+//
+//  ProfileImageService.swift
+//  ImageFeed
+//
+//  Created by Егор Свистушкин on 01.04.2023.
+//
+
+import Foundation
+
+final class ProfileImageService {
+    
+    static let shared = ProfileImageService()
+    private (set) var profileImageURL: String?
+    private let urlSession = URLSession.shared
+    private let storage = OAuth2TokenStorage()
+    private var task: URLSessionTask?
+    static let didChangeNotification = Notification.Name(rawValue: "ProfileImageProviderDidChange")
+    
+    func fetchProfileImageURL(username: String, completion: @escaping (Result<Bool, Error>) -> Void) {
+        
+        // Добавил устранение гонки, должно работать, если данные есть, то мы всё отменяем, если нет, то продолжаем запросы 4.04.23, 13:30
+        assert(Thread.isMainThread)
+        if task != nil {
+            if profileImageURL == nil {
+                task?.cancel()
+            } else {
+                return
+            }
+        } else {
+            if profileImageURL != nil {
+                return
+            }
+        }
+        
+        var request = URLRequest.makeHTTPRequest(path: "/users/\(username)", httpMethod: "GET")
+        
+        if let token = storage.token {
+            request.setToken(token)
+        }
+        
+        let task = object(for: request) { [weak self] result in
+            guard let self = self else { return }
+            switch result {
+            case .success(let data):
+                self.profileImageURL = data.profileImage.large
+                completion(.success(true))
+                NotificationCenter.default
+                    .post(name: ProfileImageService.didChangeNotification, object: self, userInfo: ["URL" : profileImageURL])
```

Comment:
Пришлось добавить, чтобы запустить проект. Давай доправим, чтобы у других ревьюверов не было проблем. В swift <5.7 внутри замыкания обращаться к свойствам класса нужно через `self`, новое API языка не всем еще доступно :)
```suggestion
                self.profileImageURL = data.profileImage.large
                completion(.success(true))
                NotificationCenter.default
                    .post(name: ProfileImageService.didChangeNotification, object: self, userInfo: ["URL" : self.profileImageURL])
```